lib/SILOptimizer/Differentiation/JVPCloner.cpp

```diff
@@ -1475,18 +1551,37 @@ void JVPCloner::Implementation::prepareForDifferentialGeneration() {
   // Initialize tangent mapping for indirect results.
   auto origIndResults = original->getIndirectResults();
   auto diffIndResults = differential.getIndirectResults();
+  auto isNonWrtInoutParameter = [&](unsigned i) {
+    auto &paramInfo = original->getLoweredFunctionType()->getParameters()[i];
+    return paramInfo.isIndirectInOut() && !getIndices().parameters->contains(i);
+  };
+  auto origNumParams = original->getLoweredFunctionType()->getNumParameters();
 #ifndef NDEBUG
-  unsigned numNonWrtInoutParameters = llvm::count_if(
-    range(original->getLoweredFunctionType()->getNumParameters()),
-    [&] (unsigned i) {
-      auto &paramInfo = original->getLoweredFunctionType()->getParameters()[i];
-      return paramInfo.isIndirectInOut() && !getIndices().parameters->contains(i);
-    });
+  unsigned numNonWrtInoutParameters =
+      llvm::count_if(range(origNumParams), isNonWrtInoutParameter);
+  assert(origIndResults.size() + numNonWrtInoutParameters ==
+         diffIndResults.size());
 #endif
-  assert(origIndResults.size() + numNonWrtInoutParameters == diffIndResults.size());
+  differentialBuilder.setInsertionPoint(differential.getEntryBlock());
+  for (auto i : range(origNumParams)) {
+    if (isNonWrtInoutParameter(i)) {
+      auto &origParam = original->getArgumentsWithoutIndirectResults()[i];
+      if (!activityInfo.isVaried(origParam, getIndices().parameters))
+        continue;
+      auto origSILType = origParam->getType();
+      auto tangentSILType = getRemappedTangentType(origSILType);
+      auto *tanBuf = differentialBuilder.createAllocStack(
+          differential.getLocation(), tangentSILType);
+      setTangentBuffer(origEntry, origParam, tanBuf);
```

Comment:
Random idea: since `setTangentBuffer` is always called along with setting `tangentBufferInitializationInfo`, how about adding a parameter to `setTangentBuffer`? That way we could avoid potential invariant errors caused by not setting `tangentBufferInitializationInfo`.

```swift
setTangentBuffer(origEntry, origParam, tanBuf, NotInitialized)
```

Reply:
Good idea, added that: now we can provide an (optional) `IsInitialized_t` to `setTangentBuffer` to set `tangentBufferInitializationInfo`.

Reply:
When would the caller let it default to `None`?

Reply:
The caller would let it default to `None` in cases when the initializedness is not known yet/will be set later.

However, there is only one place in the code like that right now in `prepareForDifferentialGeneration`:
```
...
for (auto &origBB : *original)
    for (auto i : indices(origIndResults))
      setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i]); // it is pointless to set initializedness here since we'll do it for all `diffIndResults` immediately afterwards
  for (auto &diffIndResult : diffIndResults)
    tangentBufferInitializationInfo[diffIndResult] = NotInitialized;
```

Reply:
BTW, do you understand the purpose of calling `setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i])` for every BB in `original`? Won't it be sufficient to just set it for the entry block?
```
for (auto i : indices(origIndResults))
  setTangentBuffer(differential.getEntryBlock(), origIndResults[i], diffIndResults[i]);
```

Reply:
> BTW, do you understand the purpose of calling `setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i])` for every BB in `original`? Won't it be sufficient to just set it for the entry block?

`setTangentBuffer(&origBB, origIndResults[i], diffIndResults[i])` is called for every original basic block in anticipation of future control flow support. It would sufficient to set it just for the entry block for now, but I think it's better to keep the existing logic to be future-facing.

Reply:
IMHO there isn't a great reason to allow initialization info to be `Optional` in `setTangentBuffer`, because the programming interface should in principle enforce the invariants which all call sites make assumptions about. All `VJPCloner` visitor methods plus other places currently assume that every tangent buffer has an associated initialization info. Allowing initialization info to be `Optional` feels like an ad-hoc solution which on one hand specifically caters to `prepareForDifferentialGeneration` while on the other hand making all other call sites error-prone. One can easily forget to pass initialization info to `setTangentBuffer` when adding a new visitor in the future.

Since `JVPCloner` is emitting all tangent buffers (including in `prepareForDifferentialGeneration`), maybe we can delay the call to `setTangentBuffer` to the place where we **actually know** whether it is initialized or not?